src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -679,49 +716,57 @@ private String key() {
 
     /** Handles normal build completion (including errors) but also handles the case that the flow did not even start correctly, for example due to an error in {@link FlowExecution#start}. */
     private void finish(@Nonnull Result r, @CheckForNull Throwable t) {
-        setResult(r);
-        duration = Math.max(0, System.currentTimeMillis() - getStartTimeInMillis());
-        LOGGER.log(Level.INFO, "{0} completed: {1}", new Object[] {toString(), getResult()});
-        if (listener == null) {
-            LOGGER.log(Level.WARNING, this + " failed to start", t);
-        } else {
-            RunListener.fireCompleted(WorkflowRun.this, listener);
-            if (t instanceof AbortException) {
-                listener.error(t.getMessage());
-            } else if (t instanceof FlowInterruptedException) {
-                ((FlowInterruptedException) t).handle(this, listener);
-            } else if (t != null) {
-                listener.getLogger().println(Functions.printThrowable(t).trim()); // TODO 2.43+ use Functions.printStackTrace
+        synchronized (getLogCopyGuard()) {
+            setResult(r);
+            if (completed != null) {
+                completed.set(true);
+            } else {
+                completed = new AtomicBoolean(true);
             }
-            listener.finished(getResult());
-            listener.closeQuietly();
+            duration = Math.max(0, System.currentTimeMillis() - getStartTimeInMillis());
         }
-        logsToCopy = null;
+
         try {
-            save();
-        } catch (Exception x) {
-            LOGGER.log(Level.WARNING, "failed to save " + this, x);
-        }
-        Timer.get().submit(() -> {
+            LOGGER.log(Level.INFO, "{0} completed: {1}", new Object[]{toString(), getResult()});
+            if (listener == null) {
+                // Can we actually rely on this, now that we use an initializing getter?
```

Comment:
Is this a TODO?

Reply:
Something like that, more of a "I wonder what @jglick will say in review"